apps/src/javalab/JavalabFileHelper.js

```diff
@@ -18,20 +18,80 @@ export function getDefaultFileContents(filename, viewMode) {
 export const getTabKey = index => `file-${index}`;
 
 export const fileMetadataForEditor = (sources, isEditingStartSources) => {
+  /*
+  'sources' contains file information as key-value pairs.
+  The key is the file name such as 'Class1.java'.
+  The value is a file object with the keys: text (source code),
+  tabOrder (order of the file tab in orderedTabs from 0 to (number of files - 1)),
+  isVisible, and isValidation (default is false)
+  The orderOfFiles array will contain the 'fileOrder' of each file as they are assigned
+  their tabKey names (such as 'file-0', 'file-1') which is stored in fileMetadata.
+  For example, if sources contains the following:
+  {
+    'Class1.java': {text: '', tabOrder: 2, isVisible: true},
+    'Class2.java': {text: '', tabOrder: 0, isVisible: true},
+    'Class3.java': {text: '', tabOrder: 1, isVisible: true}
+  },
+  then depending on how Object.keys iterates through the keys on sources, we could have
+  fileMetadata assigned: {'file-0': 'Class2.java', 'file-1': 'Class1.java', 'file-2': 'Class3.java'}.
+  The corresponding orderOfFiles would be [0, 2, 1], unorderedTabKeys would be ['file-0', 'file-1', 'file-2],
+  and orderedTabKeys would be ['file-0', 'file-2', 'file-1'].
+  */
   let fileMetadata = {};
   let orderedTabKeys = [];
-
-  Object.keys(sources).forEach(file => {
-    if (sources[file].isVisible || isEditingStartSources) {
-      // tabIndex should be the length of orderedTabKeys, not the index
-      // from the list, because we skip hidden files in the tabs.
-      const tabIndex = orderedTabKeys.length;
-      let tabKey = getTabKey(tabIndex);
-      fileMetadata[tabKey] = file;
-      orderedTabKeys.push(tabKey);
+  let unorderedTabKeys = [];
+  let orderOfFiles = [];
+  let fileIndex = 0;
+  let isValid = true;
+  Object.keys(sources).forEach(fileName => {
```

Comment:
Thanks! This will definitely make it more simple to construct the orderedTabKeys while possible missing tabOrders from the validation or support files. Will do!